File: pythagorean_triple_generator2.clj
;; ピタゴラス数生成2

(ns pythagorean-triple-generator2
  (:require [clojure.math.numeric-tower :refer [expt gcd]]))

(defn pythagorean-triples []
  (for [m (iterate inc 1)
        n (range 1 m)
        :when (and (== (gcd m n) 1)
                   (odd? (- m n)))]
    [(- (expt m 2) (expt n 2)) (* 2 m n) (+ (expt m 2) (expt n 2))]))

;; 利用例
(println (take 10 (pythagorean-triples)))

File: PythagoreanTripleGenerator2.hs
-- ピタゴラス数生成2

module PythagoreanTripleGenerator2 (
  pythagoreanTriples,
  main
) where

pythagoreanTriples :: Integral a => [(a, a, a)]
pythagoreanTriples =
  [(m ^ 2 - n ^ 2, 2 * m * n, m ^ 2 + n ^ 2) |
    m <- [1..], n <- [1..(m - 1)], gcd m n == 1, odd (m - n)]

-- 利用例
main :: IO ()
main = print $ take 10 pythagoreanTriples
